json-rpc/src/stream_rpc/subscription.rs

```diff
@@ -0,0 +1,256 @@
+        tokio::time::sleep(sleepy_time).await;
+    }
+}
+
+pub trait Subscription<ParamType, ReturnType>: Send + Sync + Clone + Default + 'static
+where
+    ParamType: Send + Sync + Clone + 'static,
+    ReturnType: Send + Sync + Serialize + Clone,
+{
+    fn init(&mut self, helper: &SubscriptionHelper, params: &ParamType)
+        -> Result<(), JsonRpcError>;
+    fn next(&self, helper: &SubscriptionHelper, params: &ParamType) -> Vec<ReturnType>;
+    fn on_send(&mut self, item: &ReturnType);
+
+    fn run(
+        mut self,
+        mut helper: SubscriptionHelper,
+        params: ParamType,
+    ) -> Result<JoinHandle<()>, JsonRpcError> {
+        self.init(&helper, &params)?;
+
+        Ok(tokio::spawn(async move {
+            helper.send_ok().await;
+
+            loop {
+                let items = self.next(&helper, &params);
+
+                debug!(
+                    "Client#{}: fetched {} items for {}",
+                    helper.client.id,
+                    items.len(),
+                    helper.method.as_str()
+                );
+
+                if items.is_empty() {
+                    helper.sleep_wiggled().await;
+                    continue;
+                }
+
+                helper.reset_backoff();
+
+                counters::SUBSCRIPTION_RPC_SENT
+                    .with_label_values(&[
+                        helper.client.connection_context.transport.as_str(),
+                        helper.method.as_str(),
+                        counters::RpcResult::Success.as_str(),
+                        helper.client.connection_context.sdk_info.language.as_str(),
+                        &helper
+                            .client
+                            .connection_context
+                            .sdk_info
+                            .version
+                            .to_string(),
+                    ])
+                    .inc_by(items.len() as u64);
+
+                for item in items {
+                    match helper
+                        .client
+                        .send_success(helper.jsonrpc_id.clone(), &item)
+                        .await
+                    {
+                        Err(e) => {
+                            // client has disconnected
+                            debug!("Client#{}: Send error: {:?}", &helper.client.id, e);
+                            return;
+                        }
+                        Ok(_) => self.on_send(&item),
+                    };
+                }
+            }
+        }))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        stream_rpc::{
+            connection::ConnectionContext,
+            errors::StreamError,
+            subscriptions::{SubscribeToTransactionsParams, TransactionsSubscription},
+            transport::util::Transport,
+        },
+        tests::utils::mock_db,
+    };
+    use tokio::sync::mpsc;
+
+    #[tokio::test]
+    async fn test_subscription() {
```

Comment:
This particular file and associated method `Subscription#run` is at the core of this JSON RPC PubSub feature. It's the pipeline for IOC functionality in subscriptions.rs if you will. We would benefit from some more edge cases WRT tests (ie: sleep_wiggles, is flow as expected?), or perhaps better targeting and coverage of L139-198 above. Testing it in a separate private `fn` would make that easier, esp with edge cases.

Reply:
Updated to add tests for sleep, and for non-happy path